src/Components/Components/src/RouteView.cs

```diff
@@ -84,30 +83,18 @@ private void RenderPageWithParameters(RenderTreeBuilder builder)
 
         void RenderPageCore(RenderTreeBuilder builder)
         {
-            builder.OpenComponent(0, RouteData.PageType);
-
-            foreach (var kvp in RouteData.RouteValues)
+            builder.OpenComponent<CascadingQueryValueProvider>(0);
```

Comment:
I'd love to reduce the number of implicit, system-created layers in the component hierarchy. Each layer is a nontrivial cost as we've been observing with @DamianEdwards's recent benchmarking (e.g., when we changed to have an `App` component and a `Router`).

Not suggesting this should change in this PR, but the approach I have in mind is:

 * We define a new mechanism within the cascading parameter lookup flow where, if we're scanning up the hierarchy and don't find any match for a cascading parameter, we also call some new API on the `Renderer` in case it wants to supply a value.
 * We define a new protected virtual API on `Renderer` so that subclasses can plug in their own set of `ICascadingValueSupplier` instances for this
 * We update `EndpointHtmlRenderer` and others to have implementations for form/query/etc as applicable based on the renderer type.

This will also let us remove the `CascadingModelBinder` layer.

I'm happy to do this sort of thing as a follow-up.

Reply:
> I'd love to reduce the number of implicit, system-created layers in the component hierarchy. Each layer is a nontrivial cost as we've been observing with @DamianEdwards's recent benchmarking (e.g., when we changed to have an App component and a Router).

> This will also let us remove the CascadingModelBinder layer, and means we can add as many new built-in cascaded parameter types as we want without incurring the cost of an extra layer in the component hierarchy for each one.

I would say that we shouldn't jump to conclusions with regards to perf here without more data. The data we have so far compares two very different pipelines, and we would need data with and without the component to make a proper judgement call. Otherwise, it's premature optimization.

`CascadingModelBinder` exists because the app must have the ability to influence the context in which form dispatching (and in the future, passing prerendered state from the server to the client), happens.

Reply:
> the app must have the ability to influence the context in which form dispatching

Just to be clear, I'm only talking about at the root level. Ideally there would be a built-in default behavior that does the most natural form binding, so we can reduce the number of out-of-box concepts we make people understand. People would only need to add them manually where they want to introduce a scope boundary. Is that what you're already thinking? If not, can we find a way to design the system that way?